pagopa/io-app
# chore: [IOBP-753] Add error backoff retry handling into payments domain

## Short description
This PR introduces error backoff retry management across the entire payments domain in a common manner, limiting the ability to retry too many times in the Payments section (and in a future PR for the Wallet section), preventing potential overload on the backend server. Additionally, a new component `BannerErrorState` has been added, which, in a future iteration and with @dmnplb's approval, will be moved to the `io-app-design-system` library.

## List of changes proposed in this pull request
- Added reducer, actions, state, and types about the backoff retry into payments section;
- Added a new `BannerErrorState` component following [the design of product & design](https://www.figma.com/design/4Pbt3wd9WoPnAXVMkAA6UT/App-IO-2.0---Design-System?node-id=20855-2072&t=sNmKWMXiui8WC9fp-4);
- Added the error handling into the "Payments" section to show a banner to retry the request both on the user methods carousel and user transactions list (The pull-to-refresh doesn't have any effect if both are in error, the user can retry a request only tapping on the CTA "Retry again")
- Show an error toast if the user cannot retry the request and the time is not elapsed from the previous request;
- Added a common `usePaymentsBackoffRetry` hook that handles the backoff retry in a more abstract way. It provides a `canRetryRequest` method that manages the logic behind the hood showing a countdown error toast by default.

## How to test
- Start the dev-server and the io-app in dev environment mode.
- First, check that both the onboarding and payment flows work as expected.
- Then, try closing the dev-server before tapping on the Payments section tab. You should see the banner with the retry CTA.
- When you tap the CTA to retry, verify that it follows the exponential backoff, starting from 1 second and going up to a maximum of 3 minutes.
- While the banner is visible, restart the dev-server in the background and check that the retry works as expected. (Remember, if the user retries from either banner CTA, ensure that if the request is successful, it also triggers the other request. For example, if you tap the "Retry" button for user payment methods, and the backend response is OK, the app should also retry the transactions request automatically.)

## Preview

https://github.com/user-attachments/assets/a11932ef-5202-473c-86fc-2603878663b6
